presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankBuilder.java

```diff
@@ -32,40 +32,67 @@
 
     public GroupedTopNRankBuilder(
             List<Type> sourceTypes,
             PageWithPositionComparator comparator,
+            PageWithPositionEqualsAndHash equalsAndHash,
             int topN,
-            boolean produceRowNumber,
+            boolean produceRanking,
             GroupByHash groupByHash)
     {
         this.sourceTypes = requireNonNull(sourceTypes, "sourceTypes is null");
         checkArgument(topN > 0, "topN must be > 0");
-        this.produceRowNumber = produceRowNumber;
+        this.produceRanking = produceRanking;
         this.groupByHash = requireNonNull(groupByHash, "groupByHash is null");
 
         requireNonNull(comparator, "comparator is null");
-        groupedTopNRowNumberAccumulator = new GroupedTopNRowNumberAccumulator(
-                (leftRowId, rightRowId) -> {
-                    Page leftPage = pageManager.getPage(leftRowId);
-                    int leftPosition = pageManager.getPosition(leftRowId);
-                    Page rightPage = pageManager.getPage(rightRowId);
-                    int rightPosition = pageManager.getPosition(rightRowId);
-                    return comparator.compareTo(leftPage, leftPosition, rightPage, rightPosition);
+        requireNonNull(equalsAndHash, "equalsAndHash is null");
+        groupedTopNRankAccumulator = new GroupedTopNRankAccumulator(
+                new RowComparisonStrategy()
+                {
+                    @Override
+                    public int compare(long leftRowId, long rightRowId)
+                    {
+                        Page leftPage = pageManager.getPage(leftRowId);
+                        int leftPosition = pageManager.getPosition(leftRowId);
+                        Page rightPage = pageManager.getPage(rightRowId);
+                        int rightPosition = pageManager.getPosition(rightRowId);
+                        return comparator.compareTo(leftPage, leftPosition, rightPage, rightPosition);
+                    }
+
+                    @Override
+                    public boolean equals(long leftRowId, long rightRowId)
+                    {
+                        Page leftPage = pageManager.getPage(leftRowId);
+                        int leftPosition = pageManager.getPosition(leftRowId);
+                        Page rightPage = pageManager.getPage(rightRowId);
+                        int rightPosition = pageManager.getPosition(rightRowId);
+                        return equalsAndHash.equals(leftPage, leftPosition, rightPage, rightPosition);
+                    }
+
+                    @Override
+                    public long hashCode(long rowId)
+                    {
+                        Page page = pageManager.getPage(rowId);
+                        int position = pageManager.getPosition(rowId);
+                        return equalsAndHash.hashCode(page, position);
+                    }
                 },
+                RowReferencePageManager.RESERVED_UNUSED_ROW_ID,
```

Comment:
static import, but I would just assume `-1` is the special value. I don't expect `GroupedTopNRankAccumulator` to be used in any other context.

Reply:
Actually the special is -2, since -1 is already reserved for the unknown row id